### PR TITLE
Theme Restricted Parameter

### PR DESCRIPTION
**Pull request by khacoder**
Initial PR
Issue #72

### CODE REVIEW

File: WordPress/Sniffs/Theme/DeprecatedArgsSniff.php [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+				'home'     => 'home_url()',
+				'site_url' => 'site_url()',
+			),
+		);
+
+		$types = array( T_CONSTANT_ENCAPSED_STRING );
+
+		foreach ( $checks as $key => $check ) {
+			if ( trim( $token['content'], '\"\'"' ) === $key ) {
+				$nextStackPtr = $stackPtr;
```
Comment (grappler) [inline WordPress/Sniffs/Theme/DeprecatedArgsSniff.php]: Is this line needed?

File: WordPress/Sniffs/Theme/DeprecatedArgsSniff.php [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+				$nextStackPtr = $phpcsFile->findNext( $types , $nextStackPtr + 1 );
```
Comment (grappler) [inline WordPress/Sniffs/Theme/DeprecatedArgsSniff.php]: Perhaps just place `array( T_CONSTANT_ENCAPSED_STRING )` directly in here.

File: WordPress/Sniffs/Theme/DeprecatedArgsSniff.php [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+				foreach ( $check as $key2 => $arg ) {
```
Comment (grappler) [inline WordPress/Sniffs/Theme/DeprecatedArgsSniff.php]: What about `$args_key` instead?

Comment (khacoder) [inline WordPress/Sniffs/Theme/DeprecatedArgsSniff.php]: Are numbers as bad as globals :)
Will fix in the next update.

File: WordPress/Sniffs/Theme/DeprecatedArgsSniff.php [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+				'feed_url'             => "echo esc_url( get_feed_link( 'feed' ) ), where feed is rss, rss2 or atom",
+			),
+			'get_option' => array(
+				'home'     => 'home_url()',
+				'site_url' => 'site_url()',
+			),
+		);
+
+		foreach ( $checks as $key => $check ) {
+			if ( trim( $token['content'], '\"\'"' ) === $key ) {
```
Comment (jrfnl) [inline WordPress/Sniffs/Theme/DeprecatedArgsSniff.php]: Actually, functions in PHP are case-insensitive, no matter how the function was declared and whether it is a native PHP function or a user-land function.

So any comparison for function names should be done in a case-insensitive way.

```php
function my_function() {
	// do something
}

my_function(); // works
MY_function(); // works
MY_FUNCTION(); // works
My_FuncTion(); // works
```

**Comment by jrfnl** [comment]
I suggest pausing work on this one in anticipation of the AbstractFunctionCall base class as discussed in Slack.

> 
I intend to add an AbstractFunctionCall base class to WPCS which will first process to check if something a) is a function call and b) is in the list of functions you're looking for and will only pass off the processing to the child class once that has been established

> Once in the child class, you can then use for instance the (upcoming) `getFunctionCallParameter( $offset )` method to get the exact parameter to check or if you only wanted to check whether a function is used, you could just throw an error straight away

File: WordPress/Sniffs/Theme/DeprecatedArgsSniff.php [hunk 5]
```diff
@@ -0,0 +1,359 @@
...
+					T_STRING === $tokens[ $pprev ]['code']
+				)
+			)
+			) {
+			return;
+		}
+
+		if ( isset( $this->functions[ $token['content'] ] ) ) {
+			$function      = $this->functions[ $token['content'] ];
+			$parameter_arg = $this->getFunctionCallParameter( $phpcsFile, $stackPtr, 1 );
```
Comment (jrfnl) [inline WordPress/Sniffs/Theme/DeprecatedArgsSniff.php]: Quick remark: I would make the parameter position part of the array above as for different functions, the parameter position might be different and hardcoding it to `1` here does not allow for that flexibility.

See these example sniffs which also uses this utility function:
* https://github.com/wimg/PHPCompatibility/blob/master/Sniffs/PHP/NewFunctionParametersSniff.php
* https://github.com/wimg/PHPCompatibility/blob/master/Sniffs/PHP/RemovedFunctionParametersSniff.php
* https://github.com/wimg/PHPCompatibility/blob/master/Sniffs/PHP/RequiredOptionalFunctionParametersSniff.php

The array for WP would look differently, but you can probably get some inspiration from the array as used there.

**Comment by grappler** [comment]
Not all of the parameters are actually deprecated. So I am checking if there is a reason why they were not supposed to be used. https://wordpress.slack.com/archives/core/p1479063392002489

Also it seems like `feed_url` is also not a valid parameter.

**Comment by justintadlock** [comment]
Some of the URL ones are not technically deprecated.  However, they are inconsistent.  Just look at this mess:

```
case 'stylesheet_url':
	$output = get_stylesheet_uri();
	break;
case 'stylesheet_directory':
	$output = get_stylesheet_directory_uri();
	break;
case 'template_directory':
case 'template_url':
	$output = get_template_directory_uri();
	break;
```

IIRC, once core WP added getter functions for the URLs, we decided to make them required for consistency sake.

I'd actually like to see core WP drop the entire function and separate it into multiple functions.  It might be worth exploring putting a core ticket for that together as a team.

**Comment by grappler** [comment]
Ok, I move the parameters that are not deprecated to a restricted parameters sniff so that there is a clear distinction.

> I'd actually like to see core WP drop the entire function and separate it into multiple functions. It might be worth exploring putting a core ticket for that together as a team.

Sounds like a good idea.

**Comment by pattonwebz** [comment]
Verify restrictions and alts against TRT handbook.

**Comment by jrfnl** [comment]
I will review the actual logic in the `process_parameter()` method and let you know if I have feedback.

**Comment by jrfnl** [comment]
I will also rebase the PR & update for PHPCS 3.x.

**Comment by khacoder** [comment]
I would like to finish this one if I could please.

**Comment by jrfnl** [comment]
@khacoder Saw the invite. Thanks for that. I've just (force) pushed the updated branch. You will need to reset your local branch to the remote version to get access to the changes.

As you asked to continue with the sniff yourself, I've only done a rebase and added a commit which sorts out the PHPCS 3.x compatibility and some minor updates for consistency with WPCS.

Would you like me to review the logic in the `process_parameter()` method at this time or do you want to have a look yourself first ?

**Comment by khacoder** [comment]
Sure I can take a look.

**Comment by khacoder** [comment]
Just going over the logic and reason for this sniff. It was initially set up to replicate the more-deprecated.php sniff in the theme check plugin. The theme check is currently unchanged from when this sniff was initially developed.

So why has this sniff changed? Many parameters have been added that are not deprecated, and parameters have been removed that are deprecated.

So why are Error's being generated for non deprecated parameters? I would think that they should be Warnings ?

**Comment by grappler** [comment]
As @khacoder prompted me to look closer at the PR I realised I had worked on a sniff to check for deprecated parameter values that should be merged into WPCS itself.

https://github.com/grappler/WordPress-Coding-Standards/commit/61897bf572b4d64858e8e9f22ea8e711c8ab8a95

For that reason I think this PR should be marked as wait for upstream for the moment till that sniff it added to WPCS and then this Sniff could extend that.

**Comment by jrfnl** [comment]
@grappler :+1: Sounds good to me. Go for it! (and let me know if you want me to take a look in advance)

**Comment by jrfnl** [comment]
> and then this Sniff could extend that.

@grappler Having just reviewed the upstream sniff, if you want to be able to extend that one, it will need some changes in the setup.

**Comment by justintadlock** [comment]
The team decided that this should mirror Theme Check and only catch the stylesheet and directory URLs. @khacoder can fill in the details if necessary.

Meeting: https://wordpress.slack.com/archives/C02RP4Y3K/p1528822802001090

**Comment by jrfnl** [comment]
PR #165 has been pulled which will include the WPCS deprecated parameter values sniff as created by @grappler. Once that PR has been merged, this PR can be adjusted to extend the upstream WPCS sniff.

**Pull request closed without merging**